### PR TITLE
THRIFT-4902: Swift 4 and 5 compatibility

### PR DESCRIPTION
**Pull request by Coeur**
- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?  (not required for trivial changes)
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?

This pull request:
- keeps compatibility with Swift 3
- adds support for Swift 4.0, 4.2, 5.0 and 5.1
- fixes the podspec:
  - the pod name must match the filename in current versions of CocoaPods
  - `requires_arc  = true` is already the default in CocoaPods
- `mutating func removeAtIndex` is removed because it wasn't implemented to begin with: it was just calling itself in an infinite loop.

### CODE REVIEW

**Comment by Coeur** [comment]
> The changes look okay but without cross testing how do we know they work?

From my reputation. 😉

More seriously, with CocoaPods, we have a linter, so it can build the lib for all platforms (ios/macos here) and different Swift versions. Which means that that we can validate this pull request compatibility by running:
```
pod lib lint --allow-warnings --swift-version=3
pod lib lint --allow-warnings --swift-version=4.0
pod lib lint --allow-warnings --swift-version=4.2
pod lib lint --allow-warnings --swift-version=5.0
```
I've now added a travis job for that.

**Comment by jeking3** [comment]
But if we only lint we're not checking I/O against other languages, so adding a cross test harness would be a better choice.

**Comment by Coeur** [comment]
I understand, but that's not the scope of this PR: I'm only adding compatibility with current Swift syntax here. Adding cross test support is the purpose of a separate Jira (THRIFT-4547, #1660).

And I guess that it will be easier to add tests now that the code actually builds for all Swift versions.

**Comment by jeking3** [comment]
Travis CI has Swift 4.2 installed, so if you can provide the cross test changes as a separate PR I would appreciate seeing it cross tested.  I'll watch for build completion.

File: lib/swift/Sources/TMap.swift [hunk 1]
```diff
@@ -33,12 +33,14 @@ public struct TMap<Key : TSerializable & Hashable, Value : TSerializable>: Colle
   }
   
   public mutating func updateValue(_ value: Value, forKey key: Key) -> Value? {
-    return updateValue(value, forKey: key)
+    return storage.updateValue(value, forKey: key)
   }
   
+  /*
```
Comment (kainjow) [inline lib/swift/Sources/TMap.swift]: why is this commented out?

Comment (Coeur) [reply lib/swift/Sources/TMap.swift]: Maybe, yes.

**Comment by jeking3** [comment]
Rebase on master please.  The only known build issue (after what I just merged) in on Xenial, which doesn't affect this since Swift isn't installed on that image.  Thanks.

**Comment by Coeur** [comment]
@jeking3 test passed

**Comment by arpwal** [comment]
@Coeur seems like master is broken for Swift 5. Could you help? :)

**Comment by janosvitok** [comment]
@arpwal Can you be more specific? What kind of error do you see? Does any of #2073,  #2069, #2059, #2051, #2047, #2038, #1997 help with the problem?

**Comment by Coeur** [comment]
@arpwal Sorry, I'm not available to help currently. Please check janosvitok pull requests, they are likely to help.

**Pull request merged**